examples/utils.py

```diff
@@ -242,3 +247,180 @@ def rgb_sigma_fn(t_starts, t_ends, ray_indices):
         depths.view((*rays_shape[:-1], -1)),
         extras,
     )
+
+
+@torch.no_grad()
+def render_image_with_occgrid_test(
+    max_samples: int,
+    # scene
+    radiance_field: torch.nn.Module,
+    estimator: OccGridEstimator,
+    rays: Rays,
+    # rendering options
+    near_plane: float = 0.0,
+    far_plane: float = 1e10,
+    render_step_size: float = 1e-3,
+    render_bkgd: Optional[torch.Tensor] = None,
+    cone_angle: float = 0.0,
+    alpha_thre: float = 0.0,
+    early_stop_eps: float = 1e-4,
+    # only useful for dnerf
+    timestamps: Optional[torch.Tensor] = None,
+):
+    """Render the pixels of an image."""
+    rays_shape = rays.origins.shape
+    if len(rays_shape) == 3:
+        height, width, _ = rays_shape
+        num_rays = height * width
+        rays = namedtuple_map(
+            lambda r: r.reshape([num_rays] + list(r.shape[2:])), rays
+        )
+    else:
+        num_rays, _ = rays_shape
+
+    def rgb_sigma_fn(t_starts, t_ends, ray_indices):
+        t_origins = rays.origins[ray_indices]
+        t_dirs = rays.viewdirs[ray_indices]
+        positions = (
+            t_origins + t_dirs * (t_starts[:, None] + t_ends[:, None]) / 2.0
+        )
+        if timestamps is not None:
+            # dnerf
+            t = (
+                timestamps[ray_indices]
+                if radiance_field.training
+                else timestamps.expand_as(positions[:, :1])
+            )
+            rgbs, sigmas = radiance_field(positions, t, t_dirs)
+        else:
+            rgbs, sigmas = radiance_field(positions, t_dirs)
+        return rgbs, sigmas.squeeze(-1)
+
+    N_rays = rays.origins.shape[0]
```

Comment:
Is it the same with the `num_rays` in L.279?

Reply:
Yes, I just realized that.